### DISCUSSION TITLE
Concept: Willow can directly use Matter

### DISCUSSION THREAD

**Discussion by jonsmirl** [toverainc/willow#96] [Ideas]
Willow doesn't have to go through HA, instead it can directly control Matter devices. HA can still know what is going on by subscribing to Matter events. For example if Willow is going to turn a bulb on/off then HA can subscribe to the on/off event from the bulb. Directly using Matter is going to be much faster.

To make this work, Willow will need to generate an elliptic keypair and CSR, then send that CSR over to HA. HA signs and returns the certificate chain to the Willow node. Once you do that HA and Willow will be on the same Matter fabric and can see all of the devices. To make life easier put a CASE Authenticated Tag into the certificate and then you don't need to mess with ACLs.

While you are waiting on the HA people to figure the previous paragraph out, you can commission devices directly from the Espressif code and simply ignore HA to begin with.  Use the esp-matter controller example:
https://github.com/espressif/esp-matter/tree/main/examples/controller

Espressif built a demo using this code, it already runs on the S3 Box:
https://www.youtube.com/watch?v=Jr4Lut_NgqA&ab_channel=EspressifSystems

As you get more sophisticated, HA can expose all of its automations via a Matter endpoint. Then Willow can trigger them via the Matter protocol.

The same principle would extend to a Linux based Willow inference server. Matter has full Linux support. The Willow inference server can use the same technique to directly control Matter devices.

**Comment by kristiankielhofner** [comment]
@stintel is going to love this one!

Last I looked Matter is included in ESP-IDF 5. We're currently "stuck" on ESP-IDF 4.4 because they haven't completed ESP-ADF support for ESP-IDF 5. It's in the esp-adf tree but it's rough and we have a branch tracking Willow support for ESP-ADF 2.6 with ESP-IDF 5. So that's the first issue.

The larger issue I see is the complete picture is MUCH more than speech rec "turn on light" -> Matter to device. Home Assistant (and other platforms like it) provide the user interface, command vocabulary, disambiguation, logic, context, additional integrations, scripting, etc to do what people *actually* want to do - which is go to an interface, string together a bunch of disparate and arbitrary components and integrations across thousands of platforms, scenes, etc and match it with a speech command.

That significant effort and functionality is missing in your scenario.

**Comment by jonsmirl** [comment]
There is no reason you can't do both. If the command is simple, just use Matter directly. For complex actions send them to HA. You can get good performance out of direct Matter commands - on the order of a couple hundred milliseconds.  Isn't Willow going to implement the interaction model schema?

This "which is go to an interface, string together a bunch of disparate and arbitrary components and integrations across thousands of platforms, scenes, etc" is an automation. Matter is going to introduce an automation server device API in the future. When that happens HA should move over to using that API to expose its automations. Planning for this now will let Willow interact with every Matter automation server instead of only HA.

Matter does not eliminate the need for automations, they just aren't part of the standard yet. When they arrive they will likely look like the current command invocation methods.  Name + parameters.
 
Use the Matter support in esp-matter, it is kept up to date continuously.  That is where the Espressif Matter developers work on things. The main branch needs IDF 5.0.

**Reply by kristiankielhofner** [reply]
I mentioned @stintel in this issue because he spent many hours trying to get bog-standard reference Matter components to do anything just a few days ago. @stintel please correct me if I'm wrong.

"Matter is going to introduce an automation server device API in the future"

Matter is allegedly going to do a lot in "the future". I support it and have high hopes for it but I have decades of PTSD dealing with immature and emerging ecosystems that (at least in their early forms) offer minor advantages (if any) over existing, mature, well established solutions you can actually build to and utilize without burning time and driving yourself crazy. A second spent on what I know will be "yet another Matter issue" is a second taken away from one of the other tangible and practical features we have on the roadmap.

I've seen this movie before and it's filed in the back of the horror section.

**Comment by jonsmirl** [comment]
It is fine if you want to ignore Matter for a few years and go HA only. My personal opinion is that Matter will be a strong market force a year from now. Google, Apple and Amazon aren't going to give the market a choice, they will just move everything else to legacy status and let it slowly rot away.  This is the first time the big three have gotten behind a standard as a group.

**Reply by kristiankielhofner** [reply]
As one example, I spent 15 years in VoIP. That was the EXACT same public messaging for SIP. It took half a decade for interoperability to become reality and that was only after millions of lines of compatibility code were written for every implementation in existence. Spoiler alert - it's still tricky today, more than 10 years later.

I truly hope it all comes together in the next year but like I said from my experience with SIP, I don't put any faith in the public marketing/messaging of Goliath corporations ("consortiums") who are claiming to support a standard that in many ways goes against their own vested interests.

As my Grandparents from Missouri used to say: "Don't tell me, SHOW ME".

I'm waiting for show me status - I don't want to hear about it. I want to see it.

**Comment by jonsmirl** [comment]
Google, Apple and Amazon have already updated their devices with Matter support.   Alexa, Siri, Ok Google are all working now with Matter. And you can buy Matter light bulbs on Amazon. But they sell as fast as shipments arrive so they are out of stock most of the time.
https://www.amazon.com/gp/product/B0C1JD5YXW/ref=ppx_yo_dt_b_asin_title_o02_s00?ie=UTF8&th=1
Over the next twelve months, 1,135 certified devices are in the pipeline to come to market.

**Reply by kristiankielhofner** [reply]
https://www.home-assistant.io/integrations/matter/

It's beta in HASS. They explicitly call out the immature and emerging nature of support and the protocol itself. The page is riddled with caveats such as this fun one:

"Many devices that (will) hit the market will use Thread for radio communication and Matter as a control protocol, but this is not guaranteed. For example, Thread-based devices are available that only support Apple HomeKit or some vendor-specific communication protocol. There are also a few cases where you need to apply for a (beta) firmware update on the device to enable Matter as a communication protocol. Therefore, do not assume Matter support when you see a Thread logo when looking for devices. Please be sure to look for the *Matter* logo itself (on either Wi-Fi/Ethernet-based devices or Thread) or any other confirmation by the manufacturer that the device supports Matter."

Look through a Matter implementation and these kinds of caveats are EVERYWHERE. Indicative of an emerging standard in the state I referenced from past experience.

When they support it we will automatically via them, openHAB, etc. When it's clear it's mature we can revisit this for Willow itself.

Until then I will be spending our time on this project doing what we set out to do - match and surpass privacy invading voice interfaces on the market from many of these same vendors.

I've said my piece.

**Comment by hamishcunningham** [comment]
Stacey on IoT has written about Matter quite a lot, and done some testing, e.g.:

> I am incredibly underwhelmed, especially since we’ve [seen](https://staceyoniot.com/podcast-matter-upgrades-arent-ready-for-prime-time/) [a lot](https://staceyoniot.com/when-is-apple-homekit-going-to-fix-its-broken-matter-experience/) of [misfires](https://staceyoniot.com/nanoleaf-essentials-matter-reviewed-great-but-not-quite-seamless/) so far when it comes to getting Matter devices to work on a single network with multiple devices from other ecosystems. My [one-off adding of a Matter device to a single controller](https://staceyoniot.com/tp-link-tapo-smart-plug-with-matter-simple-and-mostly-smart/) worked well, but anything more complicated breaks a little more than half the time.
https://staceyoniot.com/matter-1-1-is-here-but-its-not-a-big-deal/

Andreas Spiess' recent video looks interesting too: https://www.youtube.com/watch?v=bh-kg3fVbf8
